libraries/standard/mqtt/utest/mqtt_utest.c

```diff
@@ -78,3 +245,354 @@ void test_MQTT_Init_Invalid_params( void )
+    mqttContext.transportInterface.recv = transportRecvSuccess;
+    mqttContext.networkBuffer.size = networkBuffer.size;
+    incomingPacket.remainingLength = 2;
+    MQTT_GetIncomingPacketTypeAndLength_ExpectAnyArgsAndReturn( MQTTSuccess );
+    MQTT_GetIncomingPacketTypeAndLength_ReturnThruPtr_pIncomingPacket( &incomingPacket );
+    MQTT_DeserializeAck_ExpectAnyArgsAndReturn( MQTTBadResponse );
+    status = MQTT_Connect( &mqttContext, &connectInfo, NULL, timeout, &sessionPresent );
+    TEST_ASSERT_EQUAL_INT( MQTTBadResponse, status );
+}
+
+/* ========================================================================== */
+
+/**
+ * @brief Test that MQTT_Publish works as intended.
+ */
+void test_MQTT_Publish( void )
+{
+    MQTTContext_t mqttContext;
+    MQTTPublishInfo_t publishInfo;
+    uint16_t packetId;
+    MQTTTransportInterface_t transport;
+    MQTTFixedBuffer_t networkBuffer;
+    MQTTApplicationCallbacks_t callbacks;
+    MQTTStatus_t status;
+    size_t headerSize;
+
+    const uint16_t PACKET_ID = 1;
+
+    setupTransportInterface( &transport );
+    setupCallbacks( &callbacks );
+    setupNetworkBuffer( &networkBuffer );
+    transport.send = transportSendFailure;
+
+    memset( ( void * ) &mqttContext, 0x0, sizeof( mqttContext ) );
+    memset( ( void * ) &publishInfo, 0x0, sizeof( publishInfo ) );
+    MQTT_Init( &mqttContext, &transport, &callbacks, &networkBuffer );
+
+    /* Verify parameters. */
+    status = MQTT_Publish( NULL, &publishInfo, PACKET_ID );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+    status = MQTT_Publish( &mqttContext, NULL, PACKET_ID );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+    publishInfo.qos = MQTTQoS1;
+    status = MQTT_Publish( &mqttContext, &publishInfo, 0 );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    /* Bad Parameter when getting packet size. */
+    publishInfo.qos = MQTTQoS0;
+    MQTT_GetPublishPacketSize_ExpectAnyArgsAndReturn( MQTTBadParameter );
+    status = MQTT_Publish( &mqttContext, &publishInfo, 0 );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    /* Always return success from now on. */
+    MQTT_GetPublishPacketSize_IgnoreAndReturn( MQTTSuccess );
+
+    MQTT_SerializePublishHeader_ExpectAnyArgsAndReturn( MQTTNoMemory );
+    status = MQTT_Publish( &mqttContext, &publishInfo, 0 );
+    TEST_ASSERT_EQUAL_INT( MQTTNoMemory, status );
+
+    /* The transport interface will fail. */
+    MQTT_SerializePublishHeader_ExpectAnyArgsAndReturn( MQTTSuccess );
+    /* We need sendPacket to be called with at least 1 byte to send, so that
+     * it can return failure. This argument is the output of serializing the
+     * publish header. */
+    headerSize = 1;
+    MQTT_SerializePublishHeader_ReturnThruPtr_pHeaderSize( &headerSize );
+    status = MQTT_Publish( &mqttContext, &publishInfo, 0 );
+    TEST_ASSERT_EQUAL_INT( MQTTSendFailed, status );
+
+    /* We can ignore this now since the header size is initialized to 0, so
```

Comment:
Where is this initialized to zero? By this line `   memset( ( void * ) &publishInfo, 0x0, sizeof( publishInfo ) );`?

Reply:
This was supposed to mean that MQTT_Publish initializes the headerSize to 0, so the send for the header will be treated as success, even if `transportSendFailure` is given to function (due to the line `while( bytesRemaining > 0UL)` in the `sendPacket` function). We could explicitly give a header size through
```
MQTT_SerializePublishHeader_ExpectAnyArgsAndReturn( MQTTSuccess );
MQTT_SerializePublishHeader_ReturnThruPtr_pHeaderSize( &headerSize );
```
every time `MQTT_Publish` is called in the test, but it seemed unnecessary, so I added the call to `MQTT_SerializePublishHeader_IgnoreAndReturn( MQTTSuccess );`  instead. I updated the comment.